### DISCUSSION TITLE
Linkage error when running test-bot in a custom tap

### DISCUSSION THREAD

**Discussion by caraya** [Homebrew/discussions#2176] [Writing Formulae/Casks]
I created a custom version of a formula already in homebrew/core and am trying to push it to a custom tap I'm getting the following error. I don't know if this is caused by the formula I'm working on or is an error with the action itself. 

I have a few questions:

* Is there a way to debug the action to figure out why the error is happening
* Are the actions created in the tap repo when I run `brew tap-new` a replacement for the `brew test-bot` command?
* What are the differences between the Mac and Linux bots that make one fail but not the other? (The macos-latest test-bot runs works or fails because the Linux test-bot failed)

This is the error:

```
==> brew linkage --test caraya/rivendellweb/graphicsmagick-perl
==> FAILED
Unwanted system libraries:
  /lib/x86_64-linux-gnu/libjbig.so.0

==> brew install --only-dependencies --include-test caraya/rivendellweb/graphicsmagick-perl
==> brew test --verbose caraya/rivendellweb/graphicsmagick-perl
==> FAILED caraya/rivendellweb/graphicsmagick-perl
==> brew uninstall --force berkeley-db binutils bzip2 elfutils expat freeglut freetype gcc gdbm gettext gmp isl jasper jpeg libdrm libedit libffi libmpc libpciaccess libpng libpthread-stubs libtiff libtool libva libvdpau libx11 libxau libxcb libxdamage libxdmcp libxext libxfixes libxi libxinerama libxml2 libxrandr libxrender libxshmfence libxv libxvmc libxxf86vm little-cms2 llvm lm-sensors m4 mesa mesa-glu mpdecimal mpfr ncurses openssl@1.1 perl pkg-config python@3.9 readline sqlite unzip wayland wayland-protocols webp xinput xorgproto xz zlib zstd
Error: linkage failed
Error: 1 failed step!
brew linkage --test caraya/rivendellweb/graphicsmagick-perl
Error: Process completed with exit code 1.
```

**Comment by Bo98** [comment]
Basically what it's complaining about is it's linking with `libjbig` from the system. Since Homebrew is largely a self-contained package manager on Linux, this is a problem as it means your build is reliant on things being installed outside of Homebrew for it to work and thus is also less portable. If you need `libjbig` then you should link use Homebrew's one instead.

It's however possible that, assuming the dependency on `libjbig` is optional, you don't actually want to link with `libjpeg` at all (if, for example, you don't need its functionality) and that this may be a result of _opportunistic linking_ where the build script of the software detects `libjbig` is installed and automatically links to it. There's usually a flag you can pass to the build script to disable such behaviour.

So basically the solution is either to disable linking against this library or depend on Homebrew's version (`jbigkit`), depending on whether you think you need the dependency or not (if there is even a choice).

> Is there a way to debug the action to figure out why the error is happening

If you want to understand where the build might be picking up the library, you can check its output. This is saved in the Homebrew logs directory `~/.cache/Homebrew/Logs/<formula>`.

> Are the actions created in the tap repo when I run `brew tap-new` a replacement for the `brew test-bot` command?

The output workflow yml files invoke `brew test-bot` directly. Can you clarify what you mean?

> What are the differences between the Mac and Linux bots that make one fail but not the other? (The macos-latest test-bot runs works or fails because the Linux test-bot failed)

`jbigkit` is likely preinstalled on the system in the Linux images but not the macOS ones.

There's also differences in the linkage checker between Linux and macOS (macOS does allow system library usage for Apple-supplied libraries since there aren't different OS distributions to deal with), but the first point will be the main reason in this scenario since Apple don't provide `jbigkit`.

**Reply by caraya** [reply]
@Bo98, thanks but now I'm confused. 

> So basically the solution is either to disable linking against this library or depend on Homebrew's version (jbigkit), depending on whether you think you need the dependency or not (if there is even a choice).

The linkage error happens in the Linux test-bot, not macOS.

Here's the full run where I experienced the issue, if it helps

<https://github.com/caraya/homebrew-rivendellweb/runs/3667585408?check_suite_focus=true>

The formula doesn't use `jbigkit` at all, neither did the original formula in homebrew/core that I took adapted this one from. These are the package dependencies:

```
  depends_on "pkg-config" => :build
  depends_on "freetype"
  depends_on "jasper"
  depends_on "jpeg"
  depends_on "libpng"
  depends_on "libtiff"
  depends_on "libtool"
  depends_on "little-cms2"
  depends_on "perl"
  depends_on "webp"

  uses_from_macos "bzip2"
  uses_from_macos "libxml2"
  uses_from_macos "zlib"
```

> The output workflow yml files invoke brew test-bot directly. Can you clarify what you mean?

There is also a brew command for test-bot on my local CLI. I wasn't sure if you should use the CLI or just push the new formula and let the actions work from there.  I think I figured out that I should just push the new formula to the repo and let the actions there take care of testing and publishing since this is not an official homebrew tap.

> jbigkit is likely preinstalled on the system in the Linux images but not the macOS ones.

It doesn't appear that jbigkit is installed on ubuntu-latest I will try and add it as a dependency to see if that works as indicated by SMillerDev

**Reply by SMillerDev** [reply]
> The linkage error happens in the Linux test-bot, not macOS.

Likely because a system version to link only exists on Linux.

> The formula doesn't use jbigkit at all, neither did the original formula in homebrew/core that I took adapted this one from.

No, but the software that is being packaged does pull in the system version if it can find it. So you need to stop it from doing that, or define it as a dependency so it uses the homebrew version.

**Reply by caraya** [reply]
@SMillerDev 

I get the linkage problem. Thanks for the explanation. 

Now I'm getting an error when building the bottle on the Linux test-bot.

```
==> Starting build of caraya/rivendellweb/graphicsmagick-perl
==> brew fetch --retry caraya/rivendellweb/graphicsmagick-perl --build-bottle --force
==> brew install --only-dependencies --verbose --build-bottle caraya/rivendellweb/graphicsmagick-perl
==> brew install --verbose --build-bottle caraya/rivendellweb/graphicsmagick-perl
==> FAILED
```

Is there a way to debug this? The failure message doesn't tell me where to look for the error or provide hints on how to solve it.

**Reply by SMillerDev** [reply]
You can check the logs that are generated during the build. They should be uploaded.

**Reply by caraya** [reply]
I did but they present the same message and it doesn't tell me what might have caused the error. 

Here's the link to the log if it makes more sense to you. 

https://github.com/caraya/homebrew-rivendellweb/pull/4/checks?check_run_id=3725506112

**Reply by SMillerDev** [reply]
I posted the error in your PR. It's output into the Linux build log already, you just have to scroll back a little.

**Reply by caraya** [reply]
I figured out that the error is not with my custom formula but with a dependency that might need to be recompiled. As you said, the answer was already there. 

The problem seems to be with Jbigkit not being compiled with -fPIC so it breaks on trying to relocate objects

Instead of changing another formula I dropped support for jbig altogether and it worked just fine :D
